src/shapes/sphere.cpp

```diff
@@ -402,35 +434,64 @@ class Sphere final : public Shape<Float, Spectrum> {
             return dr::zeros<SurfaceInteraction3f>();
 
-        // Recompute ray intersection to get differentiable t
-        Float t = pi.t;
-        if constexpr (dr::is_diff_v<Float>)
-            t = dr::replace_grad(t, ray_intersect_preliminary(ray, active).t);
-
-        // TODO handle RayFlags::FollowShape and RayFlags::DetachShape
-
         // Fields requirement dependencies
-        bool need_dn_duv = has_flag(ray_flags, RayFlags::dNSdUV) ||
-                           has_flag(ray_flags, RayFlags::dNGdUV);
-        bool need_dp_duv = has_flag(ray_flags, RayFlags::dPdUV) || need_dn_duv;
-        bool need_uv     = has_flag(ray_flags, RayFlags::UV) || need_dp_duv;
+        bool need_dn_duv  = has_flag(ray_flags, RayFlags::dNSdUV) ||
+                            has_flag(ray_flags, RayFlags::dNGdUV);
+        bool need_dp_duv  = has_flag(ray_flags, RayFlags::dPdUV) || need_dn_duv;
+        bool need_uv      = has_flag(ray_flags, RayFlags::UV) || need_dp_duv;
+        bool detach_shape = has_flag(ray_flags, RayFlags::DetachShape);
+        bool follow_shape = has_flag(ray_flags, RayFlags::FollowShape);
+
+        const Point3f& center = m_center.value();
+        const Float& radius = m_radius.value();
+        const Transform4f& to_world = m_to_world.value();
+        const Transform4f& to_object = m_to_object.value();
+
+        /* If necessary, temporally suspend gradient tracking for all shape
+           parameters to construct a surface interaction completely detach from
+           the shape. */
+        dr::suspend_grad<Float> scope(detach_shape, center, radius, to_world, to_object);
 
         SurfaceInteraction3f si = dr::zeros<SurfaceInteraction3f>();
-        si.t = dr::select(active, t, dr::Infinity<Float>);
 
-        si.sh_frame.n = dr::normalize(ray(t) - m_center.value());
+        Point3f local;
+        if constexpr (IsDiff) {
+            if (follow_shape) {
+                /* FollowShape glues the interaction point with the sphere,
+                   therefore to also needs to account for a possible differential
+                   rotation in to_world. we first compute a detached intersection
+                   point in local space and transform it back in world space to
+                   get a point rigidly attached to the sphere attached point,
+                   including translation, scaling and rotation */
+                Normal3f n = dr::normalize(ray(pi.t) - center);
+                local = to_object.transform_affine(dr::fmadd(n, radius, center));
+                /* With FollowShape the local position should always be static as
+                   the intersection point follows any motion of the sphere. */
+                local = dr::detach(local);
+                si.p = to_world.transform_affine(local);
+                si.sh_frame.n = (si.p - center) / radius;
+                si.t = dr::sqrt(dr::squared_norm(si.p - ray.o) / dr::squared_norm(ray.d));
```

Comment:
Would this line leak the derivative of `ray` to `si.t` (with `FollowShape`)?

Reply:
This is fine as in the case where `ray` is moving, we care about the relative motion of the sphere and the ray (origin).